File: PyMachineInventory/views/forms/MachineForm.py
from PySide6.QtWidgets import QWidget, QSplitter, QTableWidget, QTreeWidget, QTreeWidgetItem, QVBoxLayout, QTableWidgetItem
from PySide6.QtCore import Qt

from models.AppModel import AppModel

class MachineForm(QWidget):
    def __init__(self, parent:QWidget=None):
        super().__init__(parent)
        self._model = AppModel.instance()
        self._setupUi()
    
    def _setupUi(self):
        layout = QVBoxLayout(self)
        splitter = QSplitter(Qt.Horizontal)

        # árvore esquerda
        tree = QTreeWidget()
        tree.setHeaderLabel("Categorias")
        splitter.addWidget(tree)
        
        root = QTreeWidgetItem(tree, ["Resumo do Sistema"])
        components = QTreeWidgetItem(root, ["Componentes"])
        QTreeWidgetItem(root, ["Programas"])
        QTreeWidgetItem(components, ["Rede"])
        QTreeWidgetItem(components, ["Armazenamento"])
        QTreeWidgetItem(components, ["Memória RAM"])
        
        # tabela direita
        self._table = QTableWidget(0, 2)
        self._table.setHorizontalHeaderLabels(["Item", "Valor"])
        self._table.horizontalHeader().setStretchLastSection(True)
        self._table.verticalHeader().hide()
        self._table.setEditTriggers(QTableWidget.NoEditTriggers)
        
        splitter.addWidget(self._table)
        layout.addWidget(splitter)

        # eventos
        tree.itemClicked.connect(self.on_tree_itemClicked)

        tree.expandAll()
        tree.setCurrentItem(root)
        self.on_tree_itemClicked(root, 0)

        splitter.setSizes([300, 600])

    def on_tree_itemClicked(self, item:QTreeWidgetItem, column:int):
        # Limpa a tabela atual
        self._table.setRowCount(0)
        category = item.text(0)
        machine_data = self._model.machine()

        if not machine_data:
            return

        if category == "Resumo do Sistema":
            for i in range(10):
                self._table.insertRow(i)

            self._table.setItem(0, 0, QTableWidgetItem("Nome do Sistema Operacional"))
            self._table.setItem(0, 1, QTableWidgetItem(machine_data.os))
            self._table.setItem(1, 0, QTableWidgetItem("Versão"))
            self._table.setItem(1, 1, QTableWidgetItem(machine_data.osVersion))
            self._table.setItem(2, 0, QTableWidgetItem("Arquitetura"))
            self._table.setItem(2, 1, QTableWidgetItem(machine_data.osArchitecture))
            self._table.setItem(3, 0, QTableWidgetItem("Data de Instalação"))
            self._table.setItem(3, 1, QTableWidgetItem(str(machine_data.osInstallDate)))
            self._table.setItem(4, 0, QTableWidgetItem("ID do Sistema Operacional"))
            self._table.setItem(4, 1, QTableWidgetItem(machine_data.osSerialNumber))
            self._table.setItem(5, 0, QTableWidgetItem("Organização"))
            self._table.setItem(5, 1, QTableWidgetItem(machine_data.osOrganization))
            self._table.setItem(6, 0, QTableWidgetItem("Placa Mãe"))
            self._table.setItem(6, 1, QTableWidgetItem(machine_data.motherboard))
            self._table.setItem(7, 0, QTableWidgetItem("Provedor da Placa Mãe"))
            self._table.setItem(7, 1, QTableWidgetItem(machine_data.motherboardManufacturer))
            self._table.setItem(8, 0, QTableWidgetItem("Processador"))
            self._table.setItem(8, 1, QTableWidgetItem(machine_data.processor))
            self._table.setItem(9, 0, QTableWidgetItem("Velocidade de Clock do Processador"))
            self._table.setItem(9, 1, QTableWidgetItem(f"{machine_data.processorClockSpeed / 1000:.2f} GHz"))
        
        elif category == "Armazenamento":
            data = machine_data.disks
            last_index = len(data) - 1
            row = -1

            for i, disk in enumerate(data):
                row+=1
                self._table.insertRow(row)
                self._table.setItem(row, 0, QTableWidgetItem("Nome"))
                self._table.setItem(row, 1, QTableWidgetItem(disk.name))
                
                row+=1
                self._table.insertRow(row)
                self._table.setItem(row, 0, QTableWidgetItem("Número de Série"))
                self._table.setItem(row, 1, QTableWidgetItem(disk.seriaNumber))

                row+=1
                self._table.insertRow(row)
                self._table.setItem(row, 0, QTableWidgetItem("Espaço"))
                self._table.setItem(row, 1, QTableWidgetItem(f"{int(disk.size) // pow(10, 9)} GB"))

                row+=1
                self._table.insertRow(row)
                self._table.setItem(row, 0, QTableWidgetItem("Modelo"))
                self._table.setItem(row, 1, QTableWidgetItem(disk.model))

                if i != last_index:
                    row+=1
                    self._table.insertRow(row)

        elif category == "Rede":
            data = machine_data.networkAdapters
            last_index = len(data) - 1
            row = -1

            for i, d in enumerate(data):
                row+=1
                self._table.insertRow(row)
                self._table.setItem(row, 0, QTableWidgetItem("Nome"))
                self._table.setItem(row, 1, QTableWidgetItem(d.name))
                
                row+=1
                self._table.insertRow(row)
                self._table.setItem(row, 0, QTableWidgetItem("Mac"))
                self._table.setItem(row, 1, QTableWidgetItem(d.mac))

                if i != last_index:
                    row+=1
                    self._table.insertRow(row)

        elif category == "Memória RAM":
            data = machine_data.physicalMemories
            last_index = len(data) - 1
            row = -1

            for i, d in enumerate(data):
                row+=1
                self._table.insertRow(row)
                self._table.setItem(row, 0, QTableWidgetItem("Nome"))
                self._table.setItem(row, 1, QTableWidgetItem(d.name))
                
                row+=1
                self._table.insertRow(row)
                self._table.setItem(row, 0, QTableWidgetItem("Capacidade"))
                self._table.setItem(row, 1, QTableWidgetItem(f"{int(d.capacity) // pow(10, 9)} GB"))

                row+=1
                self._table.insertRow(row)
                self._table.setItem(row, 0, QTableWidgetItem("Velocidade"))
                self._table.setItem(row, 1, QTableWidgetItem(f"{d.speed} MT/s"))

                if i != last_index:
                    row+=1
                    self._table.insertRow(row)

        elif category == "Programas":
            data = machine_data.programs
            last_index = len(data) - 1
            row = -1

            for i, d in enumerate(data):
                row+=1
                self._table.insertRow(row)
                self._table.setItem(row, 0, QTableWidgetItem("Nome"))
                self._table.setItem(row, 1, QTableWidgetItem(d.name))

                row+=1
                self._table.insertRow(row)
                self._table.setItem(row, 0, QTableWidgetItem("Versão"))
                self._table.setItem(row, 1, QTableWidgetItem(d.version))

                row+=1
                self._table.insertRow(row)
                self._table.setItem(row, 0, QTableWidgetItem("Provedor"))
                self._table.setItem(row, 1, QTableWidgetItem(d.publisher))
                
                row+=1
                self._table.insertRow(row)
                self._table.setItem(row, 0, QTableWidgetItem("Espaço"))
                self._table.setItem(row, 1, QTableWidgetItem(f"{int(d.estimatedSize) // pow(10, 3)} MB"))

                row+=1
                self._table.insertRow(row)
                self._table.setItem(row, 0, QTableWidgetItem("Apenas no Usuário Atual"))
                self._table.setItem(row, 1, QTableWidgetItem(str(d.currentUserOnly)))

                if i != last_index:
                    row+=1
                    self._table.insertRow(row)

File: application/machineform.cpp
#include "machineform.h"
#include "appmodel.h"
#include "id.h"

#include <sysinfo/sysinfo.h>
#include <wx/splitter.h>

inventory::MachineForm::MachineForm(wxWindow* parent, int border)
    : wxPanel(parent, ID_MACHINE_FORM)
    , model(AppModel::GetInstance())
    , syncTextFormat(wxString::FromUTF8("Sincronizado em %d/%m às %H:%M"))
{
    setupUI(border);
    
    Bind(wxEVT_BUTTON, &MachineForm::OnBtnSync, this, ID_MACHINE_FORM_BTN_SYNC);
    model->Bind(EVT_APPMODEL_MACHINE, &MachineForm::OnAppModelMachine, this);
}

void inventory::MachineForm::ShowMessage(const wxString& msg, int timeout)
{}

void inventory::MachineForm::BlockChanges(bool block)
{
    btnSync->Enable(!block);
}

void inventory::MachineForm::Clear()
{}

void inventory::MachineForm::UpdateTreeSelection()
{
    UpdateTreeSelection(tree->GetSelection());
}

void inventory::MachineForm::UpdateTreeSelection(wxTreeItemId id)
{
    if (!id.IsOk()) {
        return;
    }

    list->Freeze();
    list->DeleteAllItems();
	list->DeleteAllColumns();

    const wxString& label = tree->GetItemText(id);
    const wxString& lbRAM = wxString::FromUTF8("Memória RAM");
    const wxString& lbAccounts = wxString::FromUTF8("Usuários");

    if (label == "Resumo do Sistema")       setupSummary();
    else if (label == "Rede")               setupNewtwork();
    else if (label == "Armazenamento")      setupDisks();
    else if (label == lbRAM)                setupMemory();
    else if (label == "Programas")          setupPrograms();
    else if (label == lbAccounts)           setupAccounts();

    list->Thaw();
}

void inventory::MachineForm::setupUI(int border)
{
    const machine_extraction* extraction = model->GetExtraction();

    wxBoxSizer* sizer = new wxBoxSizer(wxVERTICAL);
    wxBoxSizer* footerSizer = new wxBoxSizer(wxHORIZONTAL);
    wxSplitterWindow* splitter = new wxSplitterWindow(this, wxID_ANY);
    
    tree = new wxTreeCtrl(splitter, wxID_ANY);
    tree->Bind(wxEVT_TREE_SEL_CHANGED, [this](wxTreeEvent& evt) { UpdateTreeSelection(evt.GetItem()); });

    wxTreeItemId root = tree->AddRoot(L"Resumo do Sistema");
    wxTreeItemId hardware = tree->AppendItem(root, "Componentes de Hardware");
    tree->AppendItem(hardware, "Rede");
    tree->AppendItem(hardware, "Armazenamento");
    tree->AppendItem(hardware, wxString::FromUTF8("Memória RAM"));
    
    wxTreeItemId software = tree->AppendItem(root, "Componentes de Software");
    tree->AppendItem(software, "Programas");
    tree->AppendItem(software, wxString::FromUTF8("Usuários"));

    list = new wxListCtrl(splitter, wxID_ANY, wxDefaultPosition, wxDefaultSize, wxLC_REPORT | wxLC_HRULES | wxLC_VRULES);
    

    sizer->Add(splitter, 1, wxEXPAND | wxTOP | wxRIGHT | wxLEFT, border);
    sizer->AddSpacer(6);
    sizer->Add(footerSizer, 0, wxEXPAND | wxBOTTOM | wxRIGHT | wxLEFT, border);

    const wxString& lastSyncText = extraction->datetime.Format(syncTextFormat);
    footerSizer->Add(lastSync = new wxStaticText(this, wxID_ANY, lastSyncText), 0, wxALIGN_CENTER_VERTICAL);
    footerSizer->AddStretchSpacer();
    footerSizer->Add(btnSync = new wxButton(this, ID_MACHINE_FORM_BTN_SYNC, "sincronizar"));

    splitter->SplitVertically(tree, list, 250);

    tree->Expand(root);
    tree->Expand(software);
    tree->Expand(hardware);

    if (!tree->IsSelected(root))
        tree->SelectItem(root);

    SetSizer(sizer);
}

void inventory::MachineForm::setupSummary()
{
    list->InsertColumn(0, "Item", wxLIST_FORMAT_LEFT, 200);
    list->InsertColumn(1, "Valor", wxLIST_FORMAT_LEFT, 350);
    
    long i;
    const sysinfo::machine *machine = model->GetMachine();

    i = list->InsertItem(0, "Processador");
    list->SetItem(i, 1, machine->processorName);

    i = list->InsertItem(0, "Clock do Processador");
    list->SetItem(i, 1, machine->processorClockSpeed);

    i = list->InsertItem(0, wxString::FromUTF8("Organização"));
    list->SetItem(i, 1, machine->osOrganization);

    i = list->InsertItem(0, wxString::FromUTF8("Fabricante da Placa Mãe"));
    list->SetItem(i, 1, machine->motherboardManufacturer);

    i = list->InsertItem(0, wxString::FromUTF8("Placa Mãe"));
    list->SetItem(i, 1, machine->motherboardName);

    i = list->InsertItem(0, wxString::FromUTF8("Número de Série"));
    list->SetItem(i, 1, machine->osSerialNumber);

    i = list->InsertItem(0, wxString::FromUTF8("Data de Instalação"));
    list->SetItem(i, 1, machine->osInstallDate);

    i = list->InsertItem(0, "Arquitetura");
    list->SetItem(i, 1, machine->osArchitecture);

    i = list->InsertItem(0, wxString::FromUTF8("Versão do Sistema"));
    list->SetItem(i, 1, machine->osVersion);

    i = list->InsertItem(0, "Sistema Operacional");
    list->SetItem(i, 1, machine->osName);

}

void inventory::MachineForm::setupNewtwork()
{
    list->InsertColumn(0, "Nome", wxLIST_FORMAT_LEFT, 200);
    list->InsertColumn(1, wxString::FromUTF8("Endereço MAC"), wxLIST_FORMAT_LEFT, 350);

    long i;
    const sysinfo::machine *machine = model->GetMachine();

    for (size_t j = machine->network_adapters.size(); j > 0; --j) {
        const sysinfo::network_adapter& adapter = machine->network_adapters[j - 1];

        i = list->InsertItem(0, adapter.name);
        list->SetItem(i, 1, adapter.mac);
    }
}

void inventory::MachineForm::setupDisks()
{
    list->InsertColumn(0, "Nome", wxLIST_FORMAT_LEFT, 200);
    list->InsertColumn(1, wxString::FromUTF8("Número de Série"), wxLIST_FORMAT_LEFT, 200);
    list->InsertColumn(2, wxString::FromUTF8("Espaço"), wxLIST_FORMAT_LEFT, 200);
    list->InsertColumn(3, "Modelo", wxLIST_FORMAT_LEFT, 200);

    long i;
    const sysinfo::machine *machine = model->GetMachine();

    for (size_t j = machine->disks.size(); j > 0; --j) {
        const sysinfo::disk& disk = machine->disks[j - 1];

        i = list->InsertItem(0, disk.name);
        list->SetItem(i, 1, disk.seriaNumber);
        list->SetItem(i, 2, disk.size);
        list->SetItem(i, 3, disk.model);
    }
}

void inventory::MachineForm::setupMemory()
{
    list->InsertColumn(0, "Nome", wxLIST_FORMAT_LEFT, 200);
    list->InsertColumn(1, wxString::FromUTF8("Capacidade"), wxLIST_FORMAT_LEFT, 200);
    list->InsertColumn(2, "Velocidade", wxLIST_FORMAT_LEFT, 200);

    long i;
    const sysinfo::machine *machine = model->GetMachine();

    for (size_t j = machine->physical_memories.size(); j > 0; --j) {
        const sysinfo::physical_memory& memory = machine->physical_memories[j - 1];

        i = list->InsertItem(0, memory.name);
        list->SetItem(i, 1, memory.capacity);
        list->SetItem(i, 2, memory.speed);
    }
}

void inventory::MachineForm::setupPrograms()
{
    list->InsertColumn(0, "Nome", wxLIST_FORMAT_LEFT, 200);
    list->InsertColumn(1, wxString::FromUTF8("Versão"), wxLIST_FORMAT_LEFT, 200);
    list->InsertColumn(2, "Provedor", wxLIST_FORMAT_LEFT, 200);
    list->InsertColumn(3, wxString::FromUTF8("Espaço Estimado"), wxLIST_FORMAT_LEFT, 200);
    list->InsertColumn(4, wxString::FromUTF8("Apenas no Usuário Atual"), wxLIST_FORMAT_LEFT, 200);

    long i;
    const sysinfo::machine *machine = model->GetMachine();

    for (size_t j = machine->programs.size(); j > 0; --j) {
        const sysinfo::program& program = machine->programs[j - 1];

        i = list->InsertItem(0, program.DisplayName);
        list->SetItem(i, 1, program.DisplayVersion);
        list->SetItem(i, 2, program.Publisher);
        list->SetItem(i, 3, program.EstimatedSize);
        list->SetItem(i, 4, program.CurrentUserOnly ? (wxString)"Sim" : wxString::FromUTF8("Não"));
    }
}

void inventory::MachineForm::setupAccounts()
{
    list->InsertColumn(0, "Nome", wxLIST_FORMAT_LEFT, 200);
    list->InsertColumn(1, "Nome Completo", wxLIST_FORMAT_LEFT, 200);
    list->InsertColumn(2, wxString::FromUTF8("Domínio"), wxLIST_FORMAT_LEFT, 200);
    list->InsertColumn(3, wxString::FromUTF8("Descrição"), wxLIST_FORMAT_LEFT, 200);
    //list->InsertColumn(3, "Administrador", wxLIST_FORMAT_LEFT, 200);

    long i;
    const sysinfo::machine* machine = model->GetMachine();

    for (size_t j = machine->accounts.size(); j > 0; --j) {
        const sysinfo::user_account& account = machine->accounts[j - 1];

        i = list->InsertItem(0, account.name);
        list->SetItem(i, 1, account.fullName);
        list->SetItem(i, 2, account.domain);
        list->SetItem(i, 3, account.description);
    }
}

void inventory::MachineForm::OnBtnSync(wxCommandEvent& event)
{
    BlockChanges();
    model->UpdateExtraction();
}

void inventory::MachineForm::OnAppModelMachine(wxCommandEvent& event)
{
    BlockChanges(false);

    if (!event.GetInt())
        return;
        
    const machine_extraction* extraction = (const machine_extraction*)event.GetClientData();
    lastSync->SetLabel(extraction->datetime.Format(syncTextFormat));

    UpdateTreeSelection();
}